### PR TITLE
jMAVSim: Added the option to run without the GUI

### LINKED ISSUE #59: jMAVSim without GUI
Hello,

Thanks to this post: http://discuss.px4.io/t/lightweight-or-command-line-sitl-simulation-options/2285
Modifying Visualizer3D.java3, and change setVisible(true) to 
setVisible(false).

I could run SITL using jMAVSim without the GUI as my low computational power computer was having troubles running it properly.

Do you know if most of PX4 users are using the graphics ?
Or if most of us are just wasting CPU with jMAVSim fency graphics...

I was wondering if we could set a parameter for this in order to be able to specify that we want to turn off graphics, without having to edit and compile jMAVSim .
Do you think it could be feasible and that some people could benefit from that ?

looking forward hearing from you.
Best

### PR DESCRIPTION
**Pull request by katzfey**
Added a new command line argument to allow the user to run without the GUI. The
new argument is -no-gui. This allows use of the simulator in environments that
do not have a GUI, like a headless Docker.

Fixes #59.

### CODE REVIEW

**Comment by julianoes** [comment]
Thanks for the PR @katzfey! I'll try this out and review it.

**Comment by julianoes** [comment]
@katzfey please rebase on top of master, then CI should pass.

**Comment by julianoes** [comment]
Unfortunately the problem is that the j3d libs still require a X11 DISPLAY variable as was also mentioned in https://github.com/PX4/jMAVSim/issues/59#issuecomment-421120951.

```
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.jdt.internal.jarinjarloader.JarRsrcLoader.main(JarRsrcLoader.java:61)
Caused by: java.lang.ExceptionInInitializerError
	at javax.media.j3d.GroupRetained.<init>(GroupRetained.java:155)
	at javax.media.j3d.TransformGroupRetained.<init>(TransformGroupRetained.java:116)
	at javax.media.j3d.TransformGroup.createRetained(TransformGroup.java:115)
	at javax.media.j3d.SceneGraphObject.<init>(SceneGraphObject.java:114)
	at javax.media.j3d.Node.<init>(Node.java:172)
	at javax.media.j3d.Group.<init>(Group.java:551)
	at javax.media.j3d.TransformGroup.<init>(TransformGroup.java:87)
	at me.drton.jmavsim.KinematicObject.<init>(KinematicObject.java:42)
	at me.drton.jmavsim.DynamicObject.<init>(DynamicObject.java:24)
	at me.drton.jmavsim.vehicle.AbstractVehicle.<init>(AbstractVehicle.java:26)
	at me.drton.jmavsim.vehicle.AbstractMulticopter.<init>(AbstractMulticopter.java:19)
	at me.drton.jmavsim.vehicle.Quadcopter.<init>(Quadcopter.java:33)
	at me.drton.jmavsim.Simulator.buildMulticopter(Simulator.java:431)
	at me.drton.jmavsim.Simulator.<init>(Simulator.java:312)
	at me.drton.jmavsim.Simulator.main(Simulator.java:930)
	... 5 more
Caused by: java.awt.HeadlessException: 
No X11 DISPLAY variable was set, but this program performed an operation which requires it.
	at javax.media.j3d.VirtualUniverse.<clinit>(VirtualUniverse.java:269)
	... 20 more
```

I think the change is still worthwile but it doesn't allow it to be run in CI/docker just yet.

**Comment by katzfey** [comment]
Darn! Sorry about that. You would think that if I wanted it to run in a headless Docker that I would have tested it in a headless Docker. I'll keep working on it.

**Comment by JonasVautherin** [comment]
What's the status on that? @katzfey does it run in docker now? I'll be happy to test it when it works for you!

**Comment by katzfey** [comment]
Well, I am testing in a Docker now and can reproduce the issue. I have made changes and it gets farther but still not quite there.

**Comment by katzfey** [comment]
I have it all working now on headless docker. Need to clean up some commit comments and then I can submit a new PR.

**Comment by katzfey** [comment]
@JonasVautherin Please feel free to test the work now. It seems to be working for me on a headless system with the -no-gui option. Branch is run_without_gui in my repo https://github.com/katzfey/jMAVSim

**Comment by JonasVautherin** [comment]
> Branch is run_without_gui in my repo

Which is this PR, right? :sweat_smile:

File: scripts/run_simulator.sh [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+export PX4_HOME_LAT=32.990699
+export PX4_HOME_LON=-117.128320
+export PX4_HOME_ALT=208.5
+
+ant create_run_jar copy_res
+cd out/production
+# java -XX:GCTimeRatio=20 -Djava.ext.dir= -jar jmavsim_run.jar -tcp 127.0.0.1:4560 -lockstep
+# java -XX:GCTimeRatio=20 -Djava.awt.headless=true -Djava.ext.dir= -jar jmavsim_run.jar -tcp 127.0.0.1:4560 -lockstep -no-gui
+java -XX:GCTimeRatio=20 -Djava.ext.dir= -jar jmavsim_run.jar -tcp 127.0.0.1:4560 -lockstep -no-gui
+cd ../..
```
Comment (julianoes) [inline scripts/run_simulator.sh]: Did you want to commit this one or was it just for debugging?

File: src/me/drton/jmavsim/KinematicObject.java [hunk 2]
```diff
@@ -36,16 +36,18 @@
     protected TransformGroup transformGroup;
     protected BranchGroup branchGroup;
 
-    public KinematicObject(World world) {
+    public KinematicObject(World world, boolean gui) {
```
Comment (julianoes) [inline src/me/drton/jmavsim/KinematicObject.java]: I would prefer a clearer name for `gui` like `withGui` or `showGui`.

File: scripts/run_simulator.sh (hunk 1 above)
Comment (katzfey) [reply scripts/run_simulator.sh]: Just for debugging

File: src/me/drton/jmavsim/KinematicObject.java (hunk 2 above)
Comment (katzfey) [reply src/me/drton/jmavsim/KinematicObject.java]: I'll change it to showGui

File: scripts/run_simulator.sh (hunk 1 above)
Comment (julianoes) [reply scripts/run_simulator.sh]: I would probably remove it, or otherwise clean it up and document it in the README.

**Comment by julianoes** [comment]
Awesome, thanks @katzfey!

**Pull request merged**